lib/realtime/tenants/connect/piper.ex

```diff
@@ -0,0 +1,23 @@
+defmodule Realtime.Tenants.Connect.Piper do
+  @moduledoc """
+  Pipes different commands to execute specific actions during the connection process.
+  """
+  @callback run(any()) :: {:ok, any()} | {:error, any()}
+
+  def run(pipers, init) do
+    Enum.reduce_while(pipers, {:ok, init}, fn piper, {:ok, acc} ->
+      case piper.run(acc) do
+        {:ok, result} ->
+          {:cont, {:ok, result}}
+
+        {:error, error} ->
+          {:halt, {:error, error}}
+
+        _e ->
+          {:halt, {:error, :invalid_pipe_result}}
+      end
+    end)
+  rescue
+    e -> {:error, e}
```

Comment:
were you rescuing errors prior? We prefer errors to bubble up and catch them in the super generic handlers at the very top of the app. By catching it here you'd also lose stack trace.

Reply:
👍 will remove and let it bubble up